src/lib.rs

```diff
@@ -31,15 +31,44 @@ use cfg_if::cfg_if;
 
 cfg_if! {
     if #[cfg(feature = "n-api")] {
+        #[doc(hidden)]
+        pub use neon_runtime::sys as sys;
 
         #[macro_export]
         macro_rules! register_module {
-            ($module:pat, $init:block) => {
-                #[no_mangle]
-                pub extern "C" fn neon_init_module() {
-                    println!("hello N-API, this is Neon");
-                }
-            }
+            (|$env:ident, $exports:ident| $body:block) => {
+                register_module!(env!("CARGO_PKG_NAME"), |$env, $exports| $body);
+            };
+
+            ($name:expr, |$env:ident, $exports:ident| $body:block) => {
+                extern "C" fn __neon_init_module(
+                    $env: neon::sys::napi_env,
+                    $exports: neon::sys::napi_value,
+                ) -> neon::sys::napi_value $body
+
+                #[cfg_attr(target_os = "linux", link_section = ".ctors")]
+                #[cfg_attr(target_os = "macos", link_section = "__DATA,__mod_init_func")]
+                #[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
+                pub static __register_neon_module: extern "C" fn() = {
```

Comment:
This is required to be unique, but I _think_ because of Rust name mangling that's guaranteed and we don't actually need to give it a unique name.

If we really need a unique name, we may want to take a look at the` paste` crate.